core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/index/ThreeIndexProperties.java

```diff
@@ -28,6 +30,25 @@ public <Type_> Type_ getProperty(int index) {
         }
     }
 
+    @Override
+    public <Type_> Type_ toKey(int length, int startingPosition) {
+        switch (length) {
+            case 1:
+                return toKey(startingPosition);
+            case 2:
+                return (Type_) Pair.of(toKey(startingPosition), toKey(startingPosition + 1));
```

Comment:
If there's a fail fast check in case 3 that startPosition + length < properties.length, then there should be one in case 2 too.

But see my other comment about from, to first, as that will change the impl of the fail fast.